### PR TITLE
Fix Blackholed Connection Behavior in DisruptableMockTransport

### LINKED ISSUE #61034: [CI] Reproducible Failure in CoordinatorTests.testUnresponsiveFollowerDetectedEventually
```
./gradlew ':server:test' --tests "org.elasticsearch.cluster.coordination.CoordinatorTests.testUnresponsiveFollowerDetectedEventually" -Dtests.seed=B0D0BD93A4DAC474 -Dtests.security.manager=true -Dtests.locale=mt -Dtests.timezone=Etc/GMT-4 -Druntime.java=11

```

failed in https://gradle-enterprise.elastic.co/s/gghxzf4mny3wq/tests/:server:test/org.elasticsearch.cluster.coordination.CoordinatorTests/testUnresponsiveFollowerDetectedEventually#1

with

```
  1> [2020-08-12T18:18:40,656][INFO ][o.e.c.c.CoordinatorTests ] [testUnresponsiveFollowerDetectedEventually] after test
  2> REPRODUCE WITH: ./gradlew ':server:test' --tests "org.elasticsearch.cluster.coordination.CoordinatorTests.testUnresponsiveFollowerDetectedEventually" -Dtests.seed=B0D0BD93A4DAC474 -Dtests.security.manager=true -Dtests.locale=mt -Dtests.timezone=Etc/GMT-4 -Druntime.java=11
  2> java.lang.AssertionError: node1 is a follower of node4
    Expected: is <FOLLOWER>
         but: was <CANDIDATE>
        at __randomizedtesting.SeedInfo.seed([B0D0BD93A4DAC474:D9BF60193F135F6A]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.elasticsearch.cluster.coordination.AbstractCoordinatorTestCase$Cluster.stabilise(AbstractCoordinatorTestCase.java:535)
        at org.elasticsearch.cluster.coordination.AbstractCoordinatorTestCase$Cluster.stabilise(AbstractCoordinatorTestCase.java:495)
        at org.elasticsearch.cluster.coordination.CoordinatorTests.testUnresponsiveFollowerDetectedEventually(CoordinatorTests.java:681)
  2> NOTE: leaving temporary files on disk at: /Users/armin/src/elasticsearch/server/build/testrun/test/temp/org.elasticsearch.cluster.coordination.CoordinatorTests_B0D0BD93A4DAC474-001
  2> NOTE: test params are: codec=Asserting(Lucene86), sim=Asserting(RandomSimilarity(queryNorm=false): {}), locale=mt, timezone=Etc/GMT-4
  2> NOTE: Mac OS X 10.15.6 x86_64/Oracle Corporation 11 (64-bit)/cpus=12,threads=1,free=484459008,total=536870912
  2> NOTE: All tests run in this JVM: [CoordinatorTests]
```

This only failed once ever and for the first time today it seems.
Also, this is reproducible locally, I'll take a look.

### PR DESCRIPTION
**Pull request by original-brownbear**
It is not realistic to drop messages without eventually failing.
To retain the coverage of long pauses this PR adjusts the blockholed
behavior to fail a send after 24h (which is assumed to be longer than any
timeout in the system) instead of never.

Closes #61034

### CODE REVIEW

**Review by DaveCTurner — COMMENTED** [review]
I think we also need this for responses sent in the `BLACK_HOLE` and `DISCONNECTED` states.

Does this appreciably change the running time of the tests?

**Comment by original-brownbear** [comment]
> I think we also need this for responses sent in the BLACK_HOLE and DISCONNECTED states.

I wasn't sure about adding that. It did seemed unnecessary and we currently don't do anything on disconnected either and just handle it as we handle `BLACK_HOLE`. I assumed that was because it made no functional difference because we don't have any listener or so on response sending so all it changes is add a `WARN` log in a bunch of places?

> Does this appreciably change the running time of the tests?

Not really in my testing, I bet there's some degenerate case where it does :P but over 1k+ iterations it looks irrelevant so far.

**Comment by original-brownbear** [comment]
@DaveCTurner thanks for taking a look, see here https://github.com/elastic/elasticsearch/blob/master/test/framework/src/main/java/org/elasticsearch/test/disruption/DisruptableMockTransport.java#L200 for the response handling, currently we do the same for black hole and disconnect there.

**Comment by DaveCTurner** [comment]
> I wasn't sure about adding that. It did seemed unnecessary and we currently don't do anything on disconnected either and just handle it as we handle BLACK_HOLE. I assumed that was because it made no functional difference because we don't have any listener or so on response sending so all it changes is add a WARN log in a bunch of places?

I think the removal of the join timeout will expose the same bug there, given enough iterations on CI. I.e. the join request gets through but then the connection is blackholed/disconnected before the response comes back, so it's never delivered. In reality the requester would drop the connection eventually thanks to keepalives.

**Comment by original-brownbear** [comment]
>  but then the connection is blackholed/disconnected before the response comes back, so it's never delivered

I think the reason this wasn't and isn't an issue already is that we never black-hole while we have a subset of all runnable tasks at a given timestamp when we blackhole/disconnect a connection. So the send and respond cycle will always happen in one go and it's impossible that we blackhole between the send and respond right?

Also, I'm not sure it would change any behavior for the join (or any other part of the code if we were to throw on the response sending) because it's always code like this for the response:

```
    private JoinCallback transportJoinCallback(TransportRequest request, TransportChannel channel) {
        return new JoinCallback() {

            @Override
            public void onSuccess() {
                try {
                    channel.sendResponse(Empty.INSTANCE);
                } catch (IOException e) {
                    onFailure(e);
                }
            }

            @Override
            public void onFailure(Exception e) {
                try {
                    channel.sendResponse(e);
                } catch (Exception inner) {
                    inner.addSuppressed(e);
                    logger.warn("failed to send back failure on join request", inner);
                }
            }
```

where it's just logging as a result of a failed response send.
Plus, it's kind of tricky to throw exceptions when sending responses with the way things are currently coded up because we do this:

```java
            @Override
            public void sendResponse(final TransportResponse response) {
                execute(new Runnable() {
                    @Override
                    public void run() {
```

to simulate some differences in timing when sending responses so we can't really throw to whatever code invoked `sendResponse` without losing that randomization (which makes a lot of sense since actual transport implementations will mostly fork off to an IO thread as well and won't be able to throw to the caller either) => I think we're good here in practice even though it doesn't look+feel great maybe?

**Comment by DaveCTurner** [comment]
Yeah delivering an exception to the _responder_ is kinda pointless, there's nothing it can do about it, but we should still deliver an exception response to the _requester_ in those cases.

**Comment by original-brownbear** [comment]
> but we should still deliver an exception response to the requester in those cases.

Well currently this situation isn't a thing to begin with since we always `scheduleNow` the response handling as I mentioned so we don't have to worry about this? It would just be dead code to add this handling right now wouldn't it?

**Comment by DaveCTurner** [comment]
Discussed this sync: `scheduleNow` doesn't really mean "now", we can still break the network before delivering the response.

**Comment by original-brownbear** [comment]
@DaveCTurner alright, I think https://github.com/elastic/elasticsearch/pull/61310/commits/d0b3d1fe029e6cae5d822bc0ca46febc22d49bd2 should do it here right? (ran ~20k iterations of the coordinator tests with it without issues or excessive slowness)

**Comment by original-brownbear** [comment]
urgh nervermind this needs a test adjustment now :) on it

**Comment by original-brownbear** [comment]
@DaveCTurner sorry for the noise, should be good to review now :)

File: test/framework/src/main/java/org/elasticsearch/test/disruption/DisruptableMockTransport.java [hunk 1]
```diff
@@ -198,8 +205,12 @@ public void run() {
                                 break;
 
                             case BLACK_HOLE:
+                                logger.trace("blackholed during response to {}: channel is {}", requestDescription, connectionStatus);
+                                onBlackholedDuringSend(requestId, action, destinationTransport);
+                                break;
                             case DISCONNECTED:
-                                logger.trace("dropping response to {}: channel is {}", requestDescription, connectionStatus);
+                                logger.trace("disconnected during response to {}: channel is {}", requestDescription, connectionStatus);
+                                onDisconnectedDuringSend(requestId, action, destinationTransport);
```
Comment (DaveCTurner) [inline test/framework/src/main/java/org/elasticsearch/test/disruption/DisruptableMockTransport.java:213]: Hmm I think I'd prefer a long delay on the response here using `onBlackholedDuringSend` too. We're using `DISCONNECTED` to indicate that the connection actively rejects the message, e.g. sends a RST, but if it rejects the response then the original requester is none the wiser and may wait for a long time before discovering the disconnect.

In practice it's almost never going to be that bad but I'd rather err on the pathological side if possible.

File: test/framework/src/main/java/org/elasticsearch/test/disruption/DisruptableMockTransport.java [hunk 2]
```diff
@@ -228,8 +239,14 @@ public void run() {
...
                             case BLACK_HOLE:
+                                logger.trace("blackholed during exception response to {}: channel is {}",
+                                        requestDescription, connectionStatus);
+                                onBlackholedDuringSend(requestId, action, destinationTransport);
+                                break;
                             case DISCONNECTED:
-                                logger.trace("dropping exception response to {}: channel is {}", requestDescription, connectionStatus);
+                                logger.trace("disconnected during exception response to {}: channel is {}",
+                                        requestDescription, connectionStatus);
+                                onDisconnectedDuringSend(requestId, action, destinationTransport);
```
Comment (DaveCTurner) [inline test/framework/src/main/java/org/elasticsearch/test/disruption/DisruptableMockTransport.java:249]: Similarly, we should delay notifying the sender here too.

**Review by DaveCTurner — COMMENTED** [review]
One further request about blackholed-response behaviour. I may be persuaded to keep things as they are now tho.

File: test/framework/src/main/java/org/elasticsearch/test/disruption/DisruptableMockTransport.java (hunk 1 above)
Comment (original-brownbear) [reply test/framework/src/main/java/org/elasticsearch/test/disruption/DisruptableMockTransport.java]: ++ adjusted accordingly in b181920 for both spots

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks David!

**Pull request merged**